### PR TITLE
Add Lazy property generation

### PR DESCRIPTION
**Pull request by egorikftp**
Based on https://github.com/ComposeGears/Valkyrie/pull/66 (thanks @Goooler for basic implementation)
Ref: https://github.com/ComposeGears/Valkyrie/issues/63

### CODE REVIEW

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                        config.iconPack,
+                    ).nestedClass(config.iconNestedPack)
+                }
+            }
+        }
+
+        val packageName = when {
+            config.iconNestedPack.isEmpty() -> config.iconPackage
+            else -> "${config.iconPackage}.${config.iconNestedPack.lowercase()}"
+        }
```
Comment (Goooler) [inline components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt:52]: Could be reused?

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                        imageVectorBuilderSpecs(
+                            iconName = when {
+                                config.iconNestedPack.isEmpty() -> config.iconName
+                                else -> "${config.iconNestedPack}.${config.iconName}"
+                            },
+                            vector = vector,
+                            path = {
+                                vector.nodes.forEach { node -> addVectorNode(node) }
+                            },
+                        ),
```
Comment (Goooler) [inline components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt:122]: Same?

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+import io.github.composegears.valkyrie.generator.imagevector.ImageVectorGeneratorConfig
+import io.github.composegears.valkyrie.generator.imagevector.OutputFormat.LazyDelegateProperty
+import io.github.composegears.valkyrie.generator.imagevector.backingproperty.loadIcon
+import io.github.composegears.valkyrie.parser.IconParser
+import org.junit.jupiter.api.Test
+
+class SvgGradientParserTest {
+
+    @Test
+    fun `svg linear gradient parsing`() {
```
Comment (Goooler) [inline components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt:15]: You can test these with params using `@ParameterizedTest`.

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+            import androidx.compose.ui.geometry.Offset
+            import androidx.compose.ui.graphics.Brush
+            import androidx.compose.ui.graphics.Color
+            import androidx.compose.ui.graphics.SolidColor
+            import androidx.compose.ui.graphics.vector.ImageVector
+            import androidx.compose.ui.graphics.vector.path
+            import androidx.compose.ui.unit.dp
+            import kotlin.LazyThreadSafetyMode
+
+            val LinearGradient: ImageVector by lazy(LazyThreadSafetyMode.NONE) {
```
Comment (Goooler) [inline components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt:42]: These text fixtures could be written to `resources/*.kt` and read, for shortening tests.

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt (hunk 3 above)
Comment (Goooler) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt:15]: Like

```kt
    @ParameterizedTest
    @EnumSource(value = OutputFormat::class)
    fun `svg linear gradient parsing`(outputFormat: OutputFormat) {
        val icon = loadIcon("ic_linear_gradient.svg")
        val parserOutput = IconParser.toVector(icon)
        val output = ImageVectorGenerator.convert(
            vector = parserOutput.vector,
            kotlinName = parserOutput.kotlinName,
            config = ImageVectorGeneratorConfig(
                packageName = "io.github.composegears.valkyrie.icons",
                packName = "",
                nestedPackName = "",
                outputFormat = outputFormat,
                generatePreview = false,
            ),
        ).content

        val expected = when(outputFormat) {
            OutputFormat.BackingProperty -> Path("resources/1.kt").readText()
            OutputFormat.LazyDelegateProperty -> Path("resources/2.kt").readText()
        }
        assertThat(output).isEqualTo(expected)
    }
```

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/backingproperty/BackingPropertyCommon.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+val DEFAULT_BACKING_CONFIG = ImageVectorGeneratorConfig(
+    packageName = "io.github.composegears.valkyrie.icons",
+    packName = "ValkyrieIcons",
+    nestedPackName = "",
+    outputFormat = BackingProperty,
+    generatePreview = false,
+)
+
+fun loadIcon(name: String) = Path("src/test/resources/$name")
```
Comment (Goooler) [inline components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/backingproperty/BackingPropertyCommon.kt:15]: Avoid relative path here, replace with things like https://github.com/Goooler/StringResExporter/blob/a16694448c6b88e05603f46578b676391c217236/src/test/kotlin/io/github/goooler/exporter/Utils.kt#L6-L14

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt (hunk 3 above)
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt:15]: Yes, make sense together with this suggestion https://github.com/ComposeGears/Valkyrie/pull/70#discussion_r1695038212

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt (hunk 1 above)
Comment (egorikftp) [reply components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt]: reused)

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt (hunk 2 above)
Comment (egorikftp) [reply components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt]: done

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt (hunk 3 above)
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt]: Tests now so clean, thanks for this suggestion!

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt (hunk 4 above)
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/lazyproperty/SvgGradientParserTest.kt]: Tests now very short ❤️

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/backingproperty/BackingPropertyCommon.kt (hunk 5 above)
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/backingproperty/BackingPropertyCommon.kt]: Added TestUtils

https://github.com/ComposeGears/Valkyrie/blob/213bc62bc728d92675c5609121fd980e1e52e832/components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                } else {
+                    ClassName(
+                        config.iconPackage,
+                        config.iconPack,
+                    ).nestedClass(config.iconNestedPack)
+                }
+            }
+        }
+
+        val packageName = config.resolvePackageName()
```
Comment (Goooler) [inline components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt:48]: Still duplicate.

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                                iconName = config.iconName,
+                            )
+                            else -> iconPreviewSpec(
+                                iconPackage = packageName,
+                                iconName = config.iconName,
+                            )
+                        },
+                    )
+                }
+            }
```
Comment (Goooler) [inline components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt:77]: Same above.

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt (hunk 6 above)
Comment (Goooler) [reply components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt:48]: Most logic of these 2 specs could be reused.

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+    fun loadKotlinFile(name: String): String = getResourcePath(name).readText().uniteLine()
+
+    fun getResourcePath(name: String): Path {
+        val resource = TestUtils::class.java.classLoader.getResource(name) ?: error("Resource $name not found")
+
+        return resource.toURI().toPath()
+    }
+
+    private fun String.uniteLine(): String = replace("\r\n", "\n").replace("\r", "\n")
```
Comment (Goooler) [inline components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]:
```suggestion
    fun getResourceText(name: String): String = getResourcePath(name).readLines().joinToString("\n")

    fun getResourcePath(name: String): Path {
        val resource = TestUtils::class.java.classLoader.getResource(name) ?: error("Resource $name not found")

        return resource.toURI().toPath()
    }
```

File: components/generator/imagevector/src/test/resources/output/backing/AllPathParams.kt
Comment (Goooler) [inline components/generator/imagevector/src/test/resources/output/backing/AllPathParams.kt:1]: I would like to group resource files to 

```sh
tree components/generator/imagevector/src/test/resources

components/generator/imagevector/src/test/resources
├── kt
│   ├── backing
│   │   ├── AllPathParams.kt
│   │   ├── Colored.WithoutPath.kt
│   │   ├── FillColorStroke.kt
│   │   ├── LinearGradient.kt
│   │   ├── LinearGradientWithStroke.kt
│   │   ├── OnlyPath.kt
│   │   ├── RadialGradient.kt
│   │   ├── SeveralPath.kt
│   │   ├── TransparentFillColor.kt
│   │   ├── WithoutPath.kt
│   │   ├── WithoutPath.pack.kt
│   │   ├── WithoutPath.pack.nested.preview.kt
│   │   ├── WithoutPath.pack.preview.kt
│   │   └── WithoutPath.preview.kt
│   └── lazy
│       ├── AllPathParams.kt
│       ├── Colored.WithoutPath.kt
│       ├── FillColorStroke.kt
│       ├── LinearGradient.kt
│       ├── LinearGradientWithStroke.kt
│       ├── OnlyPath.kt
│       ├── RadialGradient.kt
│       ├── SeveralPath.kt
│       ├── TransparentFillColor.kt
│       ├── WithoutPath.kt
│       ├── WithoutPath.pack.kt
│       ├── WithoutPath.pack.nested.preview.kt
│       ├── WithoutPath.pack.preview.kt
│       └── WithoutPath.preview.kt
└── xml
    ├── avd
    │   ├── ic_all_path_params.xml
    │   ├── ic_fill_color_stroke.xml
    │   ├── ic_only_path.xml
    │   ├── ic_several_path.xml
    │   ├── ic_transparent_fill_color.xml
    │   └── ic_without_path.xml
    └── svg
        ├── ic_linear_gradient.svg
        ├── ic_linear_gradient_with_stroke.svg
        └── ic_radial_gradient.svg

7 directories, 37 files
```

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt (hunk 8 above)
Comment (Goooler) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: Add we can add one more thing:

```kt
fun OutputFormat.toResourceText(
    pathToBackingProperty: String,
    pathToLazyProperty: String,
): String {
    return when (this) {
        OutputFormat.BackingProperty -> TestUtils.getResourceText(pathToBackingProperty)
        OutputFormat.LazyDelegateProperty -> TestUtils.getResourceText(pathToLazyProperty)
    }.trim()
}
```

Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: With suggested

```kotlin 
fun getResourceText(name: String): String = getResourcePath(name).readLines().joinToString("\n")
```

tests failed due to missing empty line in expected file. looks like it ignores last empty line

image

can be
```kotlin
fun getResourceText(name: String): String = Files.readString(getResourcePath(name))
```
need to test on windows machine 🙂

Comment (Goooler) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: Yeah, this is why I added `trim` to address https://github.com/ComposeGears/Valkyrie/pull/70#discussion_r1696120522.

Comment (Goooler) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: Or you can add `.trimEnd()`.

Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: trim do nothing in this case (https://github.com/ComposeGears/Valkyrie/pull/70#discussion_r1696120522)
after loading resource it already miss end new line

trim necessary in this case for output generation 😏 
```kotlin
 val expected = outputFormat.toResourceText(
            pathToBackingProperty = "output/backing/WithoutPath.preview.kt",
            pathToLazyProperty = "output/lazy/WithoutPath.preview.kt",
        )
        assertThat(output.trim()).isEqualTo(expected)
```

Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/TestUtils.kt:17]: will push working solution, see how it works

File: components/generator/imagevector/src/test/resources/output/backing/AllPathParams.kt
Comment (egorikftp) [reply components/generator/imagevector/src/test/resources/output/backing/AllPathParams.kt:1]: image

avd is animated vector drawable? but we don't have such type. Only xml/svg

maybe we can make
```
-kt /
-xml/
-svg/
```

Comment (Goooler) [reply components/generator/imagevector/src/test/resources/output/backing/AllPathParams.kt:1]: Also Android Vector Drawable? Both LGTM.

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt (hunk 7 above)
Comment (egorikftp) [reply components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt]: exctacred

File: components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt (hunk 6 above)
Comment (egorikftp) [reply components/generator/imagevector/src/main/kotlin/io/github/composegears/valkyrie/generator/imagevector/spec/BackingPropertySpec.kt]: reuse more code, other looks similar but has differences. I would like to leave readability as it just a 2 options for generation kotlin code 🙂

**Review by Goooler — APPROVED** [review]

File: components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/common/ResourceUtils.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+object ResourceUtils {
+
+    fun getResourceText(name: String): String = Files.readString(getResourcePath(name)).uniteLine()
+
+    fun getResourcePath(name: String): Path {
+        val resource = ResourceUtils::class.java.classLoader.getResource(name) ?: error("Resource $name not found")
+        return resource.toURI().toPath()
+    }
+
+    private fun String.uniteLine(): String = replace("\r\n", "\n").replace("\r", "\n")
```
[the remark this answers is not in the document]
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/common/ResourceUtils.kt:16]: No, added again due to tests fails on Windows 

https://github.com/ComposeGears/Valkyrie/actions/runs/10165587764

[the remark this answers is not in the document]
Comment (Goooler) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/common/ResourceUtils.kt:16]: It's #80.

[the remark this answers is not in the document]
Comment (egorikftp) [reply components/generator/imagevector/src/test/kotlin/io/github/composegears/valkyrie/generator/imagevector/common/ResourceUtils.kt]: Thanks, merged

**Review by Goooler — APPROVED** [review]

**Pull request merged**